Add support for artifact branches

jsenko opened the pull request:
Notes and highlights:
 - I've introduced new wrapper types for group, artifact, version and branch identifiers. I believe it makes the code more clear when a more specific type is used in the method parameter list instead of a bunch of String parameters. It also centralizes default group handling and validation. We can start using it gradually.
- Users can use branch name instead of a specific version in any REST calls where a version must be specified. I named that `versionExpression` to differentiate between cases where user has to specify plain version directly. This replaces the special keyword `latest` that we have currently. Users can use the `latest` branch, that is always available, by specifying `.../branch=latest/...`.
- As a result of supporting version expressions, I was considering restricting the format of group, artifact, version and branch identifiers, but I'm not 100% sure about this yet. It would require handling during migration process to v3.
- I've disabled some migration tests, because they import v2 migration files which do not support branches. I've been working on a migration tool that would convert these to v3 migration files, and can run before the tests, but this will happen in a separate PR.
- I also plan to introduce a use-case for branches in a next PR - automated creation of branches based on semantic versioning.

jsenko commented:
Question for reviewers: Do you think the branch feature is enough to resolve https://github.com/Apicurio/apicurio-epics/issues/83 or is another tag-like entity also needed? The difference between the concept of a branch and tag is the same as in git - branch can grow and points to a sequence of versions, while a tag is essentially just a movable pointer.

EricWittmann commented:
Why restrict groupId and artifactId to use `[a-zA-Z0-9._-]{1,512}` ?  I can see why we'd want to do that for Version, but for the other two it doesn't seem necessary and just introduces a new limitation to 3.0 that we didn't have before, with no clear benefit that I'm seeing.

jsenko commented:
> Why restrict groupId and artifactId to use `[a-zA-Z0-9._-]{1,512}` ? I can see why we'd want to do that for Version, but for the other two it doesn't seem necessary and just introduces a new limitation to 3.0 that we didn't have before, with no clear benefit that I'm seeing.

yeah, I wasn't convinced of this myself, just that now would be a good opportunity to do it. The length should remain restricted though, based on the database column size.

jsenko commented:
got it, I'll update it.

EricWittmann commented:
Can you explain in this PR, at a high level, how the branches API is designed to work?  Or how a user is expected to use branches in the "manual" use-case?  I can see how this would work in the case of automatic branches (latest + semantic versioning).

Looking at the openapi, it looks like I can do the following:
1. List all the branches for an artifact
2. List all versions on a branch of an artifact
3. Delete a branch - deletes all versions in the branch??
4. Add (existing?) version to branch

Is this right?  If yes, what's the flow for manual curation of a branch?  I assume:

1. Create a new artifact with version 1.0
2. Create a new branch called 1.x
3. Add existing 1.0 version to branch 1.x
4. Create a new version 1.1
5. Add existing 1.1 version to branch 1.x

I suppose I could add the same version to multiple branches as well, right?

EricWittmann commented on app/src/main/resources/io/apicurio/registry/storage/impl/sql/h2.ddl:
```diff
@@ -23,13 +23,17 @@ ALTER TABLE rules ADD PRIMARY KEY (groupId, artifactId, type);
 
 CREATE TABLE content (contentId BIGINT NOT NULL, canonicalHash VARCHAR(64) NOT NULL, contentHash VARCHAR(64) NOT NULL, content BYTEA NOT NULL, artifactreferences TEXT);
 ALTER TABLE content ADD PRIMARY KEY (contentId);
-ALTER TABLE content ADD CONSTRAINT UNQ_content_1 UNIQUE (contentHash);
```
Just curious - why?  :)

jsenko replied:
Just renamed `UNQ_` to `UQ_` for consistency `¯\_(ツ)_/¯`

EricWittmann replied:
Is it now UQ_ everywhere?  I didn't review that.

EricWittmann commented on app/src/main/resources/io/apicurio/registry/storage/impl/sql/h2.ddl:
```diff
@@ -40,6 +44,11 @@ CREATE HASH INDEX IDX_versions_5 ON versions(createdBy);
 CREATE INDEX IDX_versions_6 ON versions(createdOn);
 CREATE HASH INDEX IDX_versions_7 ON versions(contentId);
 
+CREATE TABLE artifact_version_branches (groupId VARCHAR(512) NOT NULL, artifactId VARCHAR(512) NOT NULL, branch VARCHAR(256) NOT NULL, branchOrder INT NOT NULL, version VARCHAR(256) NOT NULL);
```
Should it be `branchId` for consistency?

EricWittmann commented on app/src/main/resources/io/apicurio/registry/storage/impl/sql/h2.ddl:
```diff
@@ -40,6 +44,11 @@ CREATE HASH INDEX IDX_versions_5 ON versions(createdBy);
 CREATE HASH INDEX IDX_versions_7 ON versions(contentId);
 
+CREATE TABLE artifact_version_branches (groupId VARCHAR(512) NOT NULL, artifactId VARCHAR(512) NOT NULL, branch VARCHAR(256) NOT NULL, branchOrder INT NOT NULL, version VARCHAR(256) NOT NULL);
+ALTER TABLE artifact_version_branches ADD PRIMARY KEY (groupId, artifactId, branch, branchOrder);
```
Surprised to see `branchOrder` as part of the primary key.  Why?

EricWittmann replied:
Oh I see - there is no actual "branches" table - this is actually a mapping table.  This table contains all of the mappings from version to branch, but there's no actual listing of just the branches.

EricWittmann commented on app/src/main/resources/io/apicurio/registry/storage/impl/sql/mssql.ddl:
```diff
@@ -2,73 +2,85 @@
 -- *********************************************************************
 
-CREATE TABLE apicurio (prop_name VARCHAR(255) NOT NULL, prop_value VARCHAR(255));
+CREATE TABLE apicurio (prop_name NVARCHAR(255) NOT NULL, prop_value NVARCHAR(255));
```
Why this change?

jsenko replied:
NVARCHAR supports Unicode, so it's more general https://stackoverflow.com/questions/144283/what-is-the-difference-between-varchar-and-nvarchar

EricWittmann commented:
Can I add a version to a branch during version creation?  That would be interesting.

EricWittmann commented on app/src/main/java/io/apicurio/registry/rest/v3/GroupsResourceImpl.java:
```diff
@@ -934,6 +975,70 @@ public VersionMetaData createArtifactVersion(String groupId, String artifactId,
+        requireParameter("groupId", groupId);
+        requireParameter("artifactId", artifactId);
+        requireParameter("branchId", branchId);
+        requireParameter("version", data);
```
I wonder why this is called `data`...

jsenko replied:
Good catch, I'll check

EricWittmann commented on app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java:
```diff
@@ -746,108 +738,103 @@ private ArtifactMetaDataDto createArtifactWithMetadataRaw(String groupId, String
 
 
     @Override
     @Transactional
```
Are these annotations still necessary?  I seem to recall you've changed how this works so that the Tx boundary happens in the handle, no?

jsenko replied:
Not in this PR, I didn't want to do too much at once.

EricWittmann replied:
So you somehow figured out how to get around the rollback issue?  Or that's still an issue and a fix is coming in a future PR?

jsenko replied:
The issue is still present, it just shows up under specific circumstances. I'll create a reproducer test in the upcoming PR.

EricWittmann commented:
Do you have new tests for the branch API?  This is a big PR but I looked and didn't see anything.

jsenko commented:
> Can I add a version to a branch during version creation? That would be interesting.

Good idea, will try to implement this.

jsenko commented:
So, I've added the following changes based on the review:
- Remove format restrictions on group ID and artifact ID
- REST API level tests
- More consistent naming
- Improve UX by adding the following features:
  - Support creating a version and adding it to a branch (or multiple branches) in the same operation
  - Support a dedicated "rebase"-type operation, whereby the user can replace the branch with a different sequence of versions. This allows users to edit full branch history.
- Some small bug fixes

> Can you explain in this PR, at a high level, how the branches API is designed to work?

Branches represent a sequence of versions. Currently, the main use-case is that they can be referenced in endpoints where a version name is expected. Instead of specifying a version directly, users can provide a *version expression*, e.g. `branch=latest` or `branch=1.1.x`, and the version on the tip of that branch is used for the given operation. There is a special branch named `latest`, which replaces the bespoke code we have used to support referencing the latest version. This branch is protected from updates and deletions, because it is intended for use in other features, replacing the `versionOrder` column we have at the moment. There are multiple operations for users to manage branches, with the following constraints:
- Branch cannot reference a non-existing version. When a version is deleted, it is removed from all branches. Vice-versa, if no branch references a version, it is also deleted. Effectively, however, since the `latest` branch always contains all versions (curretly at least) this does not happen (but is implemented).
- Branch may reference the same version multiple times.

> Or how a user is expected to use branches in the "manual" use-case? I can see how this would work in the case of automatic branches (latest + semantic versioning)

Semantic versioning is the main use-case at the moment. The [GroupsResourceTest](https://github.com/jsenko/apicurio-registry/blob/artifact-branches-3/app/src/test/java/io/apicurio/registry/noprofile/rest/v3/GroupsResourceTest.java#L2997) provides a good showcase for available operations. Most common, I assume, will be the new feature, where users can specify the branch when creating a new version.

> I suppose I could add the same version to multiple branches as well, right?

Correct!

> There isn't an explicit Create Branch I don't think? Maybe there should be one?

I went with the *create or update*-style design, where a branch is created automatically as soon the first version is added to it. I think it fits the lightweight nature of branches, since they currently don't have a dedicated table for metadata. This can change in the future of course, e.g. if we need to track branch author, properties etc.

> What about automatic branch creation for semantic versioning? Will I be able to delete those branches? Can I change the versions available on them?

yes, the plan is that they will work as any other branch, the only difference is that versions will be added to them automatically, based on content or provided metadata.

> Can I remove a version from a branch if I added it by mistake?

yes, using the "rebase" (*create or replace*) operation.
